server/datastore/mysqlredis/hosts.go

```diff
@@ -0,0 +1,140 @@
+package mysqlredis
+
+import (
+	"context"
+	"time"
+
+	"github.com/fleetdm/fleet/v4/server/contexts/ctxerr"
+	"github.com/fleetdm/fleet/v4/server/contexts/logging"
+	"github.com/fleetdm/fleet/v4/server/datastore/redis"
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	redigo "github.com/gomodule/redigo/redis"
+)
+
+const enrolledHostsSetKey = "enrolled_hosts:host_ids"
+
+// TODO: additionally, we could have a cron job to re-sync the counts every so
+// often, it would:
+// a) SELECT COUNT(*) the hosts in the DB and compare with SCARD in redis
+// b) only if different (could even be only if "significantly" different), load
+// the IDs and replace the redis set with those.
+//
+// For additional safety, we could check the DB if we're about to reject an
+// enrollment: if so, check the COUNT(*) in the DB, and if it matches or is
+// above, reject, if it doesn't match load the IDs and replace the redis SET
+// with those.
```

Comment:
The implementation is quite straightforward for now - just track the host IDs in a Redis set whenever one is enrolled/created, and remove it when deleted (explicitly or via cleanup tasks).

What I'd suggest is to also implement the cron job to re-sync in case the redis set gets out of sync (or redis gets restarted for some reason, etc.). I would not add the "additional safety" thing I mention above, unless you feel it's crucial to be absolutely right. / @chiiph

Reply:
I would recommend implementing this now. Redis restarting should be an expected event. If it restarts at any point after any hosts get added, then we would never be able to recover ie we would only count hosts after the **last restart** of redis towards the enroll limit.

Reply:
Yes, that's what I recommend too.